api/src/main/java/org/openmrs/module/reporting/report/renderer/XlsReportRenderer.java

```diff
@@ -104,13 +105,25 @@ public void render(ReportData reportData, String argument, OutputStream out) thr
                 for (DataSetColumn column : columnList) {
 					excelBuilder.addCell(column.getLabel(), "bold,border=bottom");
                 }
-                for (DataSetRow row : dataset ) {
-					excelBuilder.nextRow();
-                    for (DataSetColumn column : columnList) {
-                    	Object cellValue = row.getColumnValue(column);
-                        excelBuilder.addCell(cellValue);
-                    }
-                }
+				if (dataset instanceof IterableSqlDataSet) {
+					DataSetRow row = ((IterableSqlDataSet) dataset).iterator().next();
+					while (row != null) {
+						excelBuilder.nextRow();
+						for (DataSetColumn column : columnList) {
+							Object cellValue = row.getColumnValue(column);
+							excelBuilder.addCell(cellValue);
+						}
+						row = ((IterableSqlDataSet) dataset).iterator().next();
```

Comment:
Instead of doing the above for each row, isn't it more efficient to do it outside the loop and store the iterator in a variable?

Reply:
Done.